AK/Optional.h

```diff
@@ -318,6 +319,58 @@ requires(!IsLvalueReference<T>) class [[nodiscard]] Optional<T> {
         }
     }
 
+    // If your function returns a clean value, that value is wrapped in a new Optional.
+    // If your function returns an ErrorOr, that error is propagated before the new Optional is returned.
+    // If your function returns an Optional, that Optional is returned as-is.
+    template<typename F,
+        typename MappedType = decltype(declval<F>()(declval<T&>())),
+        auto IsErrorOr = IsSpecializationOf<MappedType, ErrorOr>,
+        auto IsOptional = IsSpecializationOf<MappedType, Optional>,
+        typename OptionalType = Conditional<IsOptional, MappedType, Optional<ConditionallyResultType<IsErrorOr, MappedType>>>>
+    ALWAYS_INLINE Conditional<IsErrorOr, ErrorOr<OptionalType>, OptionalType> flat_map(F&& mapper)
+    {
+        if constexpr (IsErrorOr) {
+            if (m_has_value)
+                return OptionalType { TRY(mapper(value())) };
+            return OptionalType {};
+        } else if constexpr (IsOptional) {
+            if (m_has_value)
+                return mapper(value());
+            return OptionalType {};
+        } else {
+            if (m_has_value)
+                return OptionalType { mapper(value()) };
+
+            return OptionalType {};
+        }
+    }
+
+    // If your function returns a clean value, that value is wrapped in a new Optional.
+    // If your function returns an ErrorOr, that error is propagated before the new Optional is returned.
+    // If your function returns an Optional, that Optional is returned as-is.
+    template<typename F,
+        typename MappedType = decltype(declval<F>()(declval<T&>())),
```

Comment:
This should be `declval<T const&>` for the const versions. and probably have a test that you can't modify an optional through a flat_map that you only have a const& to. but that could be done later.